src/customnodesession.cpp

```diff
@@ -0,0 +1,202 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+#include "customnodesession.hpp"
+
+#include <cstdint>
+#include <functional>
+#include <utility>
+
+#include "custom_node_output_allocator.hpp"
+#include "logging.hpp"
+#include "node.hpp"
+#include "node_library.hpp"
+#include "nodeinputhandler.hpp"
+#include "pipelineeventqueue.hpp"
+
+namespace ovms {
+
+CustomNodeSession::CustomNodeSession(const NodeSessionMetadata& metadata, const std::string& nodeName, uint32_t inputsCount, const NodeLibrary& library, session_id_t shardsCount) :
+    NodeSession(metadata, nodeName, inputsCount, shardsCount),
+    library(library) {}
+
+CustomNodeSession::CustomNodeSession(const NodeSessionMetadata&& metadata, const std::string& nodeName, uint32_t inputsCount, const NodeLibrary& library, session_id_t shardsCount) :
+    NodeSession(std::move(metadata), nodeName, inputsCount, shardsCount),
+    library(library) {}
+
+CustomNodeSession::~CustomNodeSession() = default;
+
+Status CustomNodeSession::execute(PipelineEventQueue& notifyEndQueue, Node& node, const NodeLibrary& library, std::unique_ptr<struct CustomNodeParam[]>& parameters, int parametersLength) {
+    const auto& blobMap = this->inputHandler->getInputs();
+    auto inputTensorsLength = blobMap.size();
+    auto inputTensors = std::make_unique<struct CustomNodeTensor[]>(inputTensorsLength);
+    struct CustomNodeTensor* outputTensors = nullptr;
+    int outputTensorsLength = 0;
+
+    int i = 0;
+    for (const auto& [name, blob] : blobMap) {
+        inputTensors[i].name = static_cast<const char*>(name.c_str());
+        inputTensors[i].data = static_cast<uint8_t*>(blob->buffer());
+        inputTensors[i].dataLength = static_cast<uint64_t>(blob->byteSize());
+        inputTensors[i].dims = static_cast<uint64_t*>(blob->getTensorDesc().getDims().data());
+        inputTensors[i].dimsLength = static_cast<uint64_t>(blob->getTensorDesc().getDims().size());
+        inputTensors[i].precision = toCustomNodeTensorPrecision(blob->getTensorDesc().getPrecision());
+        i++;
+    }
+
+    int result = library.execute(
+        inputTensors.get(),
+        inputTensorsLength,
+        &outputTensors,
+        &outputTensorsLength,
+        parameters.get(),
+        parametersLength);
+
+    // If result is not 0, it means execution has failed.
+    // In this case shared library is responsible for cleaning up resources (memory).
+    if (result != 0) {
+        SPDLOG_LOGGER_ERROR(dag_executor_logger, "Node {}; session: {}; has failed custom node execution with return code: {}", getName(), getSessionKey(), result);
```

Comment:
This should be LOGGER_DEBUG right?

Reply:
And in all folliwing lines that does not make OVMS fail.

Reply:
After a bit of thought I think that we should have DEBUG here, and warning when the library executed just fine but we are missing information about how to cleanup the results properly.